### PR TITLE
Adds Dependency Injection to WebAPI Action Filters using Property Injection

### PR DESCRIPTION
**Pull request by ahoefling**
## Summary
This adds property injection for WebAPI filters using the current Dependency Injection model. It mimics similar techniques that are common in the Unity Dependency Injection library by creating a new `BuildUp()` extension method to manage the Property Injection.

## Why
Suppose you have an `ActionFilter` that decorates one of your WebAPI methods and you need to leverage dependency injection to access a specific service. In ASP.NET WebAPI you can't use constructor injection due to how the filters are generated in the pipeline. It is considered a best practice to use Property Injection instead of the standard constructor injection.

This can be dangerous for us to adopt into the platform as I can see property injection and the `BuildUp()` method being abused. I would like to take a closer look at the `AttributeUsage` and see if we can tightly-couple the `[Dependency]` attribute just to `ActionFilter`. **I do not want and do not think we should support property injection in DNN, this is a discussion I have had with @mitchelsellers and I am not trying to bring it back up with this PR.** I am using property injection here as it appears to be required to get Dependency Injection support in ActionFilters

## Example
```c#
public class MyCustomFilter : ActionFilterAttribute
{
    [Dependency]
    public IMyCustomService MyCustomService { get; set; }

    public override void OnActionExecuted(HttpActionExecutedContext actionExecutedContext)
    {
        // When this method executes MyCustomService will be injected and ready for use
    }
}
```

## Steps to Merge
| Task | Status |
|------|-------|
| Unit Tests | ✔ Success |
| Test Installer | ✔ Success |
| XML Documentation | ✔ Success |

### CODE REVIEW

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:13]: One possible thought here, not that this is a good thing...but a thought.  Should this be moved as an internal to the implementation in the action filter area to prevent mis-use?

File: DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs
Comment (mitchelsellers) [inline DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs:27]: While we are in here can we remove the commented out code

**Review by mitchelsellers — CHANGES_REQUESTED** [review]
Overall, I really like this idea as it helps with ActionFilters to use the newer API's and constructs.  But I am VERY concerned about attempted misuse.  However, by only implementing BuildUp in the code for ActionFilters we protect by proxy from misuse.

.NET Does not provide a method for us to limit attribute usage to a specific type however.

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:13]: This is something I was thinking, but it does not work without making a few other changes. The Web API code is contained in the `DotNetNuke.Web` project, which means we need to add `[InternalsVisibleTo]` assembly declaration in the Dependency Injection project. 

I think this would be an acceptable solution

File: DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs:27]: Yes we can, no problem

**Comment by ahoefling** [comment]
@mitchelsellers Thanks for the review, do you have any opinions on performance implications of this PR?

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (mitchelsellers) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:13]: Honestly, I believe that is a great solution for this, and I'm sure other things that we need to do, but only for internal benefit.  it would prevent us from just making everything public all the time

**Comment by mitchelsellers** [comment]
> @mitchelsellers Thanks for the review, do you have any opinions on performance implications of this PR?

At this point in time, nothing concrete necessarily.  There would be some impact, but traditionally Action Filters will not have many members/properties so it shouldn't be too big of a deal.

I think I'd want to try and profile this one

**Comment by ahoefling** [comment]
My biggest concern is action filters that run before or during a request. If I remember correctly from some enterprise projects I worked on years ago a blocking thread through cause performance issues with the `ApiController`. 

@mitchelsellers 
> I think I'd want to try and profile this one

I agree and I think we should test the following items prior to accepting this PR
- Performance Profile
- DotNetBenchmarking Results (if they are easy to obtain)
- Memory Leak Analysis - I don't want another DI memory leak to come up when we should be able to load test it prior to merging

Are we in agreement that we should move forward and turn this PR into a mergable PR?

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (bdukes) [inline DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:16]:
```suggestion
        /// Injects property dependency for public or protected
```

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (bdukes) [inline DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:30]: It might be worth profiling whether caching would help here.  If we're doing this many times per request, looking up the properties to fill might become a hot path to avoid.

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (bdukes) [inline DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:21]: Do we want this method to take an `ActionFilter`, to better document the expected usage?  We'd need to be explicit to create an overload if there was another type we wanted to support.

File: DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs:20]: Is DI going to resolve the global container or a scoped container here?  Do we need to think about scoping?  I would hope it would automatically pull the scoped container, which should already be in use, but want to make sure we don't make the wrong assumption here.

**Review by bdukes — APPROVED** [review]

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:21]: I think that is a great idea. As far as I understand all WebAPI `ActionFilter` objects implement the `IFilter` interface. It should be safe to type check on that.

I would like to extend this code for MVC filters as well let's focus on WebAPI for now

Comment (mitchelsellers) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:21]: I really like the IFilter or similar limitation here.  Great idea @bdukes

File: DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs:20]: As far as I understand it will resolve the scoped provider. @mitchelsellers do you know more about this?

This is the technically correct way to access the `IServiceProvider` instead of using the `Globals.DependencyProvider`. I thought since we can register the `DnnActionFilterProvider` let's just resolve the `IServiceProvider` in the constructor

Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs:20]: How do we verify this? Do we have an easy way where we can just attach a debugger and check the `IServiceProvider`?

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:30]: Won't caching the objects cause side-effects if 2 different users make the same request? Wouldn't user 1 get the scoped service from user 2 or vice-versa?

Comment (bdukes) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:30]: I'm thinking about caching the property lookup, not the object itself.

```c#
var properties = GetDependencyProperties(myObject.GetType());

private IEnumerable<PropertyInfo> GetDependencyProperties(Type type)
{
    return CBO.GetCachedObject<IEnumerable<PropertyInfo>(
        new CacheItemArgs($"DependencyProperties_{type.TypeFullName}"),
        (args) => 
            type
                .GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic)
                .Where(propertyInfo => propertyInfo.GetCustomAttribute<DependencyAttribute>() != null)
                .ToList());
}
```

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs:13]: This needs to be moved into `DotNetNuke.Web` because I don't want `DotNetNuke.DependencyInjection` to take a dependency on `System.Web`. 

The new plan will be keeping the exact same namespace but move it into the `DotNetNuke.Web` project. This is a common technique I've seen in various DI projects across .NET

Comment (ahoefling) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs]: I'm going to resolve this conversation since I have moved the code. Please open a new one if you think there needs more discussion

File: DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/Api/Internal/DnnActionFilterProvider.cs]: Removed stale code in latest update

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs]: I have added CBO caching in the latest code push

**Comment by ahoefling** [comment]
I made changes as requested in the initial review and fixed the failing unit tests. This should generate a build for us to take a closer look at and see how and if this all works. After investigating the unit tests I think we can add quite a few unit tests around this new code to ensure it is thoroughly tested. I would like to do that prior to merging it

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:48]: I'm removing the `?.` from the `Where` call b/c [`GetProperties`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getproperties?view=netframework-4.8#System_Type_GetProperties_System_Reflection_BindingFlags_) should never return `null`.

I'm also adding `ToList` b/c we want to cache the results of the query.  Otherwise we're just caching the query, and it'll go back to the data source each time it's called.

```suggestion
                        .Where(propertyInfo => propertyInfo.GetCustomAttribute<DependencyAttribute>() != null)
                        .ToList(),
```

**Review by bdukes — APPROVED** [review]
Looks great, thanks!

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:48]: This makes sense, thank you for the explanation

File: DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.DependencyInjection/Extensions/BuildUpExtensions.cs]: This has been updated in the latest code push to use `IFilter`. I'm going to resolve this conversation

**Comment by bdukes** [comment]
@ahoefling is our current status that we're waiting on unit tests for this one?  Should I mark this as On Hold until those come in?

**Comment by ahoefling** [comment]
Correct we are waiting on a few things for this PR
- Unit Tests
- Sample Module
- Test PR installer

I haven't had a chance to work on those yet. If we want to get this in for 9.8.0 I can find time tonight

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (ahoefling) [inline DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:35]: This is incorrect, it will return the PropertyInfo as the type and not the correct property type. This needs to be updated to

```suggestion
                var service = container.GetService(property.PropertyType);
```

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (ahoefling) [inline DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:38]: The parameters on this are backwards

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (ahoefling) [inline DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:51]: We should update this to check for properties that have an accessor (set method). If this is not true we will most likely see runtime exceptions.

I would expect the code snippet below to throw an exception during the buildup

```c#
public IEventLogger Logger { get; }
```

Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs]: This has been fixed in the latest code push

**Comment by ahoefling** [comment]
I have updated the accessor of the BuildUp API to internal and added unit tests.

If there is no other feedback on this, I think it is good to go

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:24]:
```suggestion
        /// Injects property dependency for properties that are decorated with <see cref="DependencyAttribute"/>.
```

File: DNN Platform/DotNetNuke.Web/Startup.cs
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/Startup.cs:20]: Does it make sense for this to be a singleton, since it's only manually accessed once and then stored in global configuration?

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (bdukes) [inline DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:60]: Every time I look at this I'm confused by this `false`
```suggestion
                    saveInDictionary: false);
```

File: DNN Platform/Tests/DotNetNuke.Tests.Web/DependencyInjection/BuildUpExtensionsTests.cs
Comment (bdukes) [inline DNN Platform/Tests/DotNetNuke.Tests.Web/DependencyInjection/BuildUpExtensionsTests.cs:154]: We don't need to duplicate the logic from the production code in order to test with the cache, we just need to make sure the callback gets used.  I guess this can be moved into the `SetUp`, since it no longer depends on the type being retrieved.
```suggestion
            var mockCbo = new Mock<ICBO>();
            mockCbo
                .Setup(x => x.GetCachedObject<IEnumerable<PropertyInfo>>(It.IsAny<CacheItemArgs>(), It.IsAny<CacheItemExpiredCallback>(), It.IsAny<bool>()))
                .Returns<CacheItemArgs, CacheItemExpiredCallback, bool>((args, callback, _) => (IEnumerable<PropertyInfo>)callback(args));

            CBO.SetTestableInstance(mockCbo.Object);
        }

```

**Review by bdukes — APPROVED** [review]
A couple of little things, but overall approach looks good.  I believe this will be in 9.8.1 or 9.9.0, since 9.8.0 RC has already started, FYI.

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:24]: makes sense to me, thanks!

File: DNN Platform/DotNetNuke.Web/Startup.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/Startup.cs:20]: If it is only instantiated once and then stored in the global configuration I think it makes sense to keep it as a transient. Otherwise it would store that singleton instance in the Dependency Injection container. If you think the `IFilterProvider` will be needed outside the global configuration then I agree we should make it a singleton.

File: DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs
Comment (ahoefling) [reply DNN Platform/DotNetNuke.Web/DependencyInjection/BuildUpExtensions.cs:59]: this certainly makes the code more readable. Let's do it!

File: DNN Platform/Tests/DotNetNuke.Tests.Web/DependencyInjection/BuildUpExtensionsTests.cs
Comment (ahoefling) [reply DNN Platform/Tests/DotNetNuke.Tests.Web/DependencyInjection/BuildUpExtensionsTests.cs:154]: 🤦‍♂️ I knew there was a way to do this but couldn't figure it out. Thanks for the tips, I'll update the unit tests

**Comment by ahoefling** [comment]
@bdukes 
> I believe this will be in 9.8.1 or 9.9.0, since 9.8.0 RC has already started

Thanks for the update on when this will be merged. If the maintainers are happy with how this PR is looking I am thinking of starting a 2nd PR for MVC filters.

File: DNN Platform/Tests/DotNetNuke.Tests.Web/Api/DnnActionFilterProviderTests.cs
Comment (bdukes) [inline DNN Platform/Tests/DotNetNuke.Tests.Web/Api/DnnActionFilterProviderTests.cs:31]: This `SetUp` could probably also use a pass-through logic, rather than hardcoding the return value.  Here's a generic setup for `ICBO.GetCachedObject`
```suggestion
            var mockCache = new Mock<ICBO>();
            mockCache
                .Setup(x => x.GetCachedObject<It.IsAnyType>(It.IsAny<CacheItemArgs>(), It.IsAny<CacheItemExpiredCallback>(), It.IsAny<bool>()))
                .Returns(new InvocationFunc(invocation => ((CacheItemExpiredCallback)invocation.Arguments[1])((CacheItemArgs)invocation.Arguments[0])));
```

Comment (bdukes) [reply DNN Platform/Tests/DotNetNuke.Tests.Web/Api/DnnActionFilterProviderTests.cs:31]: This should probably be put in a test helper class, since it's useful in many places.

**Comment by bdukes** [comment]
:+1: go ahead on the PR for MVC filters

**Review by mitchelsellers — APPROVED** [review]

File: DNN Platform/Tests/DotNetNuke.Tests.Web/DependencyInjection/BuildUpExtensionsTests.cs
Comment (bdukes) [inline DNN Platform/Tests/DotNetNuke.Tests.Web/DependencyInjection/BuildUpExtensionsTests.cs:139]:
```suggestion
        private void VerifyEventLoggerInvoked() => this.VerifyEventLoggerInvoked(Times.Once());

        private void VerifyEventLoggerInvoked(Times times) =>
```

File: DNN Platform/DotNetNuke.Web/Api/Internal/ServicesRoutingManager.cs
Comment (valadas) [inline DNN Platform/DotNetNuke.Web/Api/Internal/ServicesRoutingManager.cs:26]: Looks like there is a duplicate using here

**Pull request merged**
